### PR TITLE
nrf_rpc: Modify the nRF RPC transport to use the RPMsg Service

### PR DESCRIPTION
**Pull request by maje-emb**
Port the nRF RPC "LL RPMsg" transport to use the "RPMsg Service"

samples: nrf_rpc: entropy_nrf53: Modify to use new version of nRF RPC.

Ref: NCSDK-9241

Signed-off-by: Marcin Jeliński <marcin.jelinski@nordicsemi.no>

### CODE REVIEW

File: samples/nrf_rpc/entropy_nrf53/cpuapp/prj.conf [hunk 1]
```diff
@@ -9,10 +9,12 @@ CONFIG_HEAP_MEM_POOL_SIZE=4096
 
 CONFIG_OPENAMP=y
```
Comment (kapi-no) [inline samples/nrf_rpc/entropy_nrf53/cpuapp/prj.conf:10]: do you need to enable Open AMP now or does it automatically is enabled when enabling the RPMsg service:

CONFIG_RPMSG_SERVICE=y

File: subsys/nrf_rpc/nrf_rpc_rpmsg.c [hunk 2]
```diff
@@ -85,45 +91,86 @@ static void ll_event_handler(struct rp_ll_endpoint *endpoint,
...
+#if MASTER
+	NRF_RPC_INF("RPC master");
+	while (!rpmsg_service_endpoint_is_bound(endpoint_id)) {
+		k_sleep(K_MSEC(1));
+	}
+
+	NRF_RPC_DBG("RPC master sem take");
+	rpmsg_service_send(endpoint_id, (uint8_t *)"", 0);
+	k_sem_take(&master_handshake_sem, K_FOREVER);
+#endif
```
Comment (kapi-no) [inline subsys/nrf_rpc/nrf_rpc_rpmsg.c:151]: You can use #if #else construct for this and put the RPC master block first as it easier to read not negated definitions.

**Review by kapi-no — CHANGES_REQUESTED** [review]

File: samples/nrf_rpc/entropy_nrf53/cpuapp/prj.conf (hunk 1 above)
Comment (maje-emb) [reply samples/nrf_rpc/entropy_nrf53/cpuapp/prj.conf]: Yes, it is not necessary to enable Open AMP directly.

File: subsys/nrf_rpc/nrf_rpc_rpmsg.c [hunk 3]
```diff
@@ -85,45 +91,86 @@ static void ll_event_handler(struct rp_ll_endpoint *endpoint,
...
 }
 
-int nrf_rpc_tr_init(nrf_rpc_tr_receive_handler_t callback)
+static int endpoint_cb(struct rpmsg_endpoint *ept, void *data, size_t len,
+	uint32_t src, void *priv)
 {
-	int err = 0;
+	if (len == 0) {
+		if (!is_handshake_done) {
+#if !MASTER
```
[the remark this answers is not in the document]
Comment (maje-emb) [reply subsys/nrf_rpc/nrf_rpc_rpmsg.c]: Updated following the advice. The compiler should make the appropriate optimization.

File: subsys/nrf_rpc/nrf_rpc_rpmsg.c (hunk 2 above)
Comment (maje-emb) [reply subsys/nrf_rpc/nrf_rpc_rpmsg.c]: Updated following the advice.

**Review by KAGA164 — APPROVED** [review]

File: subsys/nrf_rpc/nrf_rpc_rpmsg.c [hunk 4]
```diff
@@ -85,45 +89,83 @@ static void ll_event_handler(struct rp_ll_endpoint *endpoint,
...
 
-error_exit:
-	NRF_RPC_DBG("nRF RPC Initialization error %d", err);
+int nrf_rpc_tr_init(nrf_rpc_tr_receive_handler_t callback)
+{
+	NRF_RPC_ASSERT(callback != NULL);
+	receive_callback = callback;
 
-	return translate_error(err);
+	if (IS_ENABLED(CONFIG_RPMSG_SERVICE_MODE_MASTER)) {
```
Comment (grochu) [inline subsys/nrf_rpc/nrf_rpc_rpmsg.c:134]: Just a suggestion: consider conditional compilation instead of the `if/else` statement.

**Review by grochu — COMMENTED** [review]
Looks good, I just posted a single suggestion.

File: subsys/nrf_rpc/nrf_rpc_rpmsg.c (hunk 4 above)
Comment (KAGA164) [reply subsys/nrf_rpc/nrf_rpc_rpmsg.c:134]: Why do you prefer conditional compilation here ?

Comment (grochu) [reply subsys/nrf_rpc/nrf_rpc_rpmsg.c:134]: To remove the non relevant part from compilation, not a big deal though, as both blocks are small.

Comment (maje-emb) [reply subsys/nrf_rpc/nrf_rpc_rpmsg.c:134]: This will be optimized at the building stage. The `if/else` statement contains a constant value.

Comment (grochu) [reply subsys/nrf_rpc/nrf_rpc_rpmsg.c:134]: Alright, makes sense.

**Review by grochu — APPROVED** [review]

**Review by kapi-no — APPROVED** [review]

**Pull request merged**